commit/chainfee/outcome.go

```diff
@@ -0,0 +1,185 @@
+	aggObs := AggregateObservation{
+		FeeComponents:     make(map[cciptypes.ChainSelector][]types.ChainFeeComponents),
+		NativeTokenPrices: make(map[cciptypes.ChainSelector][]cciptypes.BigInt),
+		FChain:            make(map[cciptypes.ChainSelector][]int),
+		ChainFeeUpdates:   make(map[cciptypes.ChainSelector][]plugincommon.ChainFeeUpdate),
+		Timestamps:        []time.Time{},
+	}
+
+	for _, ao := range aos {
+		obs := ao.Observation
+
+		// FeeComponents
+		for chainSel, feeComp := range obs.FeeComponents {
+			aggObs.FeeComponents[chainSel] = append(aggObs.FeeComponents[chainSel], feeComp)
+		}
+
+		// NativeTokenPrices
+		for chainSel, tokenPrice := range obs.NativeTokenPrices {
+			aggObs.NativeTokenPrices[chainSel] = append(aggObs.NativeTokenPrices[chainSel], tokenPrice)
+		}
+
+		// FChain
+		for chainSel, f := range obs.FChain {
+			aggObs.FChain[chainSel] = append(aggObs.FChain[chainSel], f)
+		}
+
+		for chainSel, feeUpdate := range obs.ChainFeeUpdates {
+			aggObs.ChainFeeUpdates[chainSel] = append(aggObs.ChainFeeUpdates[chainSel], feeUpdate)
+		}
+
+		// Timestamps
+		aggObs.Timestamps = append(aggObs.Timestamps, obs.Timestamp)
+	}
+
+	return aggObs
+}
+
+// getGasPricesToUpdate checks which chain fees need to be updated based on the observed chain fee prices and
+// the fee quoter updates.
+// A chain fee is selected for update if it meets one of 2 conditions:
+// 1. If time passed since the last update is greater than the stale threshold.
+// 2. If deviation between the fee quoter and feed exceeds the chain's configured threshold.
+func (p *processor) getGasPricesToUpdate(
+	currentChainUSDFees map[cciptypes.ChainSelector]plugincommon.ChainFeeUSDPrices,
+	latestUpdates map[cciptypes.ChainSelector]plugincommon.ChainFeeUpdate,
+	obsTimestamp time.Time,
+) []cciptypes.GasPriceChain {
+	var gasPrices []cciptypes.GasPriceChain
+	feeInfo := p.cfg.FeeInfo
+
+	for chain, currentChainFee := range currentChainUSDFees {
+		packedFee := cciptypes.NewBigInt(currentChainFee.ToPackedFee())
+		lastUpdate, exists := latestUpdates[chain]
+		nextUpdateTime := lastUpdate.Timestamp.Add(p.cfg.RemoteGasPriceBatchWriteFrequency.Duration())
+		// If the chain is not in the fee quoter updates or is stale, then we should update it
+		if !exists || obsTimestamp.After(nextUpdateTime) {
+			gasPrices = append(gasPrices, cciptypes.GasPriceChain{
+				ChainSel: chain,
+				GasPrice: packedFee,
+			})
+			continue
+		}
+
+		if feeInfo == nil {
+			continue
+		}
+		ci, ok := feeInfo[chain]
+		if !ok {
+			p.lggr.Warnf("could not find fee info for chain %d", chain)
+			continue
+		}
+
+		// Checks if executionFee or dataAvFee deviates from the last update
+		if mathslib.Deviates(
+			currentChainFee.ExecutionFeePriceUSD,
+			lastUpdate.ChainFee.ExecutionFeePriceUSD,
+			ci.ExecDeviationPPB.Int64()) ||
+			mathslib.Deviates(
+				currentChainFee.DataAvFeePriceUSD,
+				lastUpdate.ChainFee.DataAvFeePriceUSD,
+				ci.DataAvDeviationPPB.Int64()) {
```

Comment:
nit: If the if condition is more than one line its better to break out the subconditions into vars IMO